RecoLocalMuon/GEMSegment/plugins/GEMSegmentAlgorithm.cc

```diff
@@ -0,0 +1,474 @@
+      }
+	  
+      // calculate cut criteria for simple running mean distance cut:
+      //dXclus = fabs(running_meanX[NNN] - running_meanX[MMM]);
+      //dYclus = fabs(running_meanY[NNN] - running_meanY[MMM]);
+      // calculate minmal distance between precluster boxes containing the hits:
+      if ( running_meanX[NNN] > running_meanX[MMM] ) dXclus_box = seed_minX[NNN] - seed_maxX[MMM];
+      else                                           dXclus_box = seed_minX[MMM] - seed_maxX[NNN];
+      if ( running_meanY[NNN] > running_meanY[MMM] ) dYclus_box = seed_minY[NNN] - seed_maxY[MMM];
+      else                                           dYclus_box = seed_minY[MMM] - seed_maxY[NNN];
+	  
+	  
+      if( dXclus_box < dXclusBoxMax && dYclus_box < dYclusBoxMax ) {
+	// merge clusters!
+	// merge by adding seed NNN to seed MMM and erasing seed NNN
+	    
+	// calculate running mean for the merged seed:
+	if(seeds[NNN].size()+seeds[MMM].size() != 0) {
+	  running_meanX[MMM] = (running_meanX[NNN]*seeds[NNN].size() + running_meanX[MMM]*seeds[MMM].size()) / (seeds[NNN].size()+seeds[MMM].size());
+	  running_meanY[MMM] = (running_meanY[NNN]*seeds[NNN].size() + running_meanY[MMM]*seeds[MMM].size()) / (seeds[NNN].size()+seeds[MMM].size());
+	}
+
+	// update min/max X and Y for box containing the hits in the merged cluster:
+	if ( seed_minX[NNN] <  seed_minX[MMM] ) seed_minX[MMM] = seed_minX[NNN];
+	if ( seed_maxX[NNN] >  seed_maxX[MMM] ) seed_maxX[MMM] = seed_maxX[NNN];
+	if ( seed_minY[NNN] <  seed_minY[MMM] ) seed_minY[MMM] = seed_minY[NNN];
+	if ( seed_maxY[NNN] >  seed_maxY[MMM] ) seed_maxY[MMM] = seed_maxY[NNN];
+	    
+	// add seed NNN to MMM (lower to larger number)
+	seeds[MMM].insert(seeds[MMM].end(),seeds[NNN].begin(),seeds[NNN].end());
+	    
+	// mark seed NNN as used (at the moment just set running mean to 999999.)
+	running_meanX[NNN] = running_max;
+	running_meanY[NNN] = running_max;
+	// we have merged a seed (NNN) to the highter seed (MMM) - need to contimue to 
+	// next seed (NNN+1)
+	break;
+      }
+    }
+  }
+
+  // hand over the final seeds to the output
+  // would be more elegant if we could do the above step with 
+  // erasing the merged ones, rather than the 
+  for(size_t NNN = 0; NNN < seeds.size(); ++NNN) {
+    if(running_meanX[NNN] == running_max) continue; //skip seeds that have been marked as used up in merging
+    rechits_clusters.push_back(seeds[NNN]);
+  }
+
+  return rechits_clusters; 
+}
+
+
+GEMSegmentAlgorithm::ProtoSegments 
+GEMSegmentAlgorithm::chainHits(const EnsembleHitContainer& rechits) {
+
+  ProtoSegments rechits_chains; 
+  EnsembleHitContainer temp;
+  ProtoSegments seeds;
+
+  std::vector <bool> usedCluster;
+
+  // split rechits into subvectors and return vector of vectors:
+  // Loop over rechits
+  // Create one seed per hit
+  for(unsigned int i = 0; i < rechits.size(); ++i) {
+    temp.clear();
+    temp.push_back(rechits[i]);
+    seeds.push_back(temp);
```

Comment:
how about
ProtoSegments seeds();
seeds.reserve(rechits.size());
std::vector<bool> usedCluster(rechits.size(),false);
for ( unsigned int i=0; i<rechits.size(); ++i)
    seeds.push_back(EnsembleHitContainer(1,rechits[i]));